File: DevExpress/XPF/DevExpress.Xpf.DemoBase/DevExpress.Xpf.DemoBase/DemosHelpers/DataClasses.cs
#region Copyright (c) 2000-2024 Developer Express Inc.
/*
{*******************************************************************}
{                                                                   }
{       Developer Express .NET Component Library                    }
{                                                                   }
{                                                                   }
{       Copyright (c) 2000-2024 Developer Express Inc.              }
{       ALL RIGHTS RESERVED                                         }
{                                                                   }
{   The entire contents of this file is protected by U.S. and       }
{   International Copyright Laws. Unauthorized reproduction,        }
{   reverse-engineering, and distribution of all or any portion of  }
{   the code contained in this file is strictly prohibited and may  }
{   result in severe civil and criminal penalties and will be       }
{   prosecuted to the maximum extent possible under the law.        }
{                                                                   }
{   RESTRICTIONS                                                    }
{                                                                   }
{   THIS SOURCE CODE AND ALL RESULTING INTERMEDIATE FILES           }
{   ARE CONFIDENTIAL AND PROPRIETARY TRADE                          }
{   SECRETS OF DEVELOPER EXPRESS INC. THE REGISTERED DEVELOPER IS   }
{   LICENSED TO DISTRIBUTE THE PRODUCT AND ALL ACCOMPANYING .NET    }
{   CONTROLS AS PART OF AN EXECUTABLE PROGRAM ONLY.                 }
{                                                                   }
{   THE SOURCE CODE CONTAINED WITHIN THIS FILE AND ALL RELATED      }
{   FILES OR ANY PORTION OF ITS CONTENTS SHALL AT NO TIME BE        }
{   COPIED, TRANSFERRED, SOLD, DISTRIBUTED, OR OTHERWISE MADE       }
{   AVAILABLE TO OTHER INDIVIDUALS WITHOUT EXPRESS WRITTEN CONSENT  }
{   AND PERMISSION FROM DEVELOPER EXPRESS INC.                      }
{                                                                   }
{   CONSULT THE END USER LICENSE AGREEMENT FOR INFORMATION ON       }
{   ADDITIONAL RESTRICTIONS.                                        }
{                                                                   }
{*******************************************************************}
*/
#endregion Copyright (c) 2000-2024 Developer Express Inc.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Windows.Media;
using System.Xml.Serialization;
using DevExpress.DemoData.Helpers;
using DevExpress.Utils;
namespace DevExpress.Xpf.DemoBase.DataClasses {
	public class Movie {
		public int ID { get; set; }
		public string Name { get; set; }
		public DateTime OrderDate { get; set; }
		public decimal Price { get; set; }
		public int Quantity { get; set; }
		public decimal Total { get { return Price * Quantity; } }
		public string Preview { get; set; }
	}
	public class Cars {
		ImageSource imageSource;
		public int ID { get; set; }
		public string Trademark { get; set; }
		public string Model { get; set; }
		public int HP { get; set; }
		public double Liter { get; set; }
		public int Cyl { get; set; }
		[XmlElement("Transmiss Speed Count")]
		public int TransmissSpeedCount { get; set; }
		[XmlElement("Transmiss Automatic")]
		public string TransmissAutomatic { get; set; }
		public int MPGCity { get; set; }
		public int MPGHighway { get; set; }
		public string Category { get; set; }
		public string Description { get; set; }
		public string Hyperlink { get; set; }
		public byte[] Picture { get; set; }
		public ImageSource ImageSource {
			get {
				if(imageSource == null)
					imageSource = ImageSourceHelper.GetImageSource(new MemoryStream(Picture));
				return imageSource;
			}
		}
		public decimal Price { get; set; }
		[XmlElement("Delivery Date")]
		public DateTime DeliveryDate { get; set; }
		[XmlElement("Is In Stock")]
		public bool IsInStock { get; set; }
	}
	public class Order {
		public int ID { get; set; }
		public string Name { get; set; }
		public DateTime OrderDate { get; set; }
		public decimal Price { get; set; }
		public int Quantity { get; set; }
		public decimal Total { get { return Price * Quantity; } }
		public string Preview { get; set; }
	}
	public class Employee : IComparable, INotifyPropertyChanged {
		public int Id { get; set; }
		public int ParentId { get; set; }
		public string FirstName { get; set; }
		public string MiddleName { get; set; }
		public string LastName { get; set; }
		public string JobTitle { get; set; }
		public string Phone { get; set; }
		public string EmailAddress { get; set; }
		public string AddressLine1 { get; set; }
		public string City { get; set; }
		public string StateProvinceName { get; set; }
		public string PostalCode { get; set; }
		public string CountryRegionName { get; set; }
		[XmlIgnore]
		string groupNameCore;
		public string GroupName {
			get { return groupNameCore; }
			set {
				if(groupNameCore != value) {
					RaisePropertyChanged(nameof(GroupName));
					groupNameCore = value;
				}
			}
		}
		public DateTime BirthDate { get; set; }
		public DateTime HireDate { get; set; }
		public string Gender { get; set; }
		public string MaritalStatus { get; set; }
		public string Title { get; set; }
		public byte[] ImageData { get; set; }
		public ImageSource Image { get { return Gender == "F" ? EmployeesData.ImageFemale : EmployeesData.ImageMale; } }
		public Uri SvgImage { get { return GetSvgImageUri(); } }
		public override string ToString() {
			return FirstName + " " + LastName;
		}
		#region Equality
		public int CompareTo(object obj) {
			Employee empl = obj as Employee;
			if(empl == null)
				throw new ArgumentException();
			return string.Compare(FirstName + LastName, empl.FirstName + empl.LastName);
		}
		#endregion
		public override bool Equals(object obj) {
			if(obj == null) return false;
			return ToString() == obj.ToString();
		}
		public override int GetHashCode() {
			return Id;
		}
		void RaisePropertyChanged(string propertyName) {
			if(PropertyChanged != null) {
				PropertyChangedEventArgs e = new PropertyChangedEventArgs(propertyName);
				PropertyChanged(this, e);
			}
		}
		Uri GetSvgImageUri() {
			if(Gender == "M")
				return EmployeesData.PersonImages["Mr"];
			if(MaritalStatus == "M")
				return EmployeesData.PersonImages["Mrs"];
			if(MaritalStatus == "S")
				return EmployeesData.PersonImages["Miss"];
			return EmployeesData.PersonImages["Ms"];
		}
		public event PropertyChangedEventHandler PropertyChanged;
	}
	public class OrderDataSourceCreator {
		public static List<Order> CreateDataSource() { return CreateDataSource(1000); }
		public static List<Order> CreateDataSource(int rowCount) {
			List<Order> list = new List<Order>();
			for(int i = 0; i < rowCount; i++) {
				list.Add(CreateItem(i + 1));
			}
			return list;
		}
		static Order CreateItem(int id) {
			Order order = new Order();
			order.ID = id;
			order.Name = "Name " + id.ToString();
			order.OrderDate = DateTime.Today.AddDays(-id % 20);
			order.Price = 100 + id % 10;
			order.Quantity = 10 + id % 15;
			order.Preview = "c:/0.png";
			return order;
		}
	}
	public class MovieDataSourceCreator {
		public static List<Movie> CreateDataSource() { return CreateDataSource(10); }
		public static List<Movie> CreateDataSource(int rowCount) {
			List<Movie> list = new List<Movie>();
			for(int i = 0; i < 1000; i++) {
				list.Add(CreateItem(i + 1));
			}
			return list;
		}
		static Movie CreateItem(int id) {
			Movie order = new Movie();
			order.ID = id;
			order.Name = "Name " + id.ToString();
			order.OrderDate = DateTime.Today.AddDays(-id % 20);
			order.Price = 100 + id % 10;
			order.Quantity = 10 + id % 15;
			order.Preview = "movies/0.wmv";
			return order;
		}
	}
	[XmlRoot("NewDataSet")]
	public class CarsData : List<Cars> {
#pragma warning disable DX0008 // data for demos is manually checked
		static readonly XmlSerializer Serializer = new Native.Serialization.GeneratedSerializers.CarsData.CarsDataSerializer();
		public static List<Cars> DataSource { get { return (List<Cars>)Serializer.Deserialize(GetDataStream()); } }
#pragma warning restore DX0008
		static Stream GetDataStream() { return AssemblyHelper.GetResourceStream(typeof(CarsData).Assembly, "Data/Cars.xml", true); }
	}
	[XmlRoot("Employees")]
	public class EmployeesData : List<Employee> {
#pragma warning disable DX0008 // data for demos is manually checked
		static readonly XmlSerializer Serializer = new Native.Serialization.GeneratedSerializers.EmployeesData.EmployeesDataSerializer();
		public static List<Employee> DataSource { get { return (List<Employee>)Serializer.Deserialize(GetDataStream()); } }
#pragma warning restore DX0008
		static Stream GetDataStream() { return AssemblyHelper.GetResourceStream(typeof(EmployeesData).Assembly, "Data/Employees.xml", true); }
		static readonly Lazy<ImageSource> imageMale = new Lazy<ImageSource>(() => ImageSourceHelper.GetImageSource(AssemblyHelper.GetResourceUri(typeof(EmployeesData).Assembly, "DemosHelpers/Images/Person_Male.png")));
		internal static ImageSource ImageMale { get { return imageMale.Value; } }
		static readonly Lazy<ImageSource> imageFemale = new Lazy<ImageSource>(() => ImageSourceHelper.GetImageSource(AssemblyHelper.GetResourceUri(typeof(EmployeesData).Assembly, "DemosHelpers/Images/Person_Female.png")));
		internal static ImageSource ImageFemale { get { return imageFemale.Value; } }
		static Dictionary<string, Uri> personImages = CreatePersonImages();
		internal static Dictionary<string, Uri> PersonImages { get { return personImages; } }
		static Dictionary<string, Uri> CreatePersonImages() {
			var result = new Dictionary<string, Uri>();
			result.Add("Mr", GetPersonImage("Mr"));
			result.Add("Miss", GetPersonImage("Miss"));
			result.Add("Mrs", GetPersonImage("Mrs"));
			result.Add("Ms", GetPersonImage("Ms"));
			return result;
		}
		static Uri GetPersonImage(string person) {
			return AssemblyHelper.GetResourceUri(typeof(EmployeesData).Assembly, string.Format("DemosHelpers/Images/Person_{0}.svg", person));
		}
	}
#pragma warning disable DX0008 // data for demos is manually checked
	[XmlRoot("Employees")]
	public class EmployeesWithPhotoData : List<Employee> {
		static readonly XmlSerializer Serializer = new Native.Serialization.GeneratedSerializers.EmployeesWithPhotoData.EmployeesWithPhotoDataSerializer();
		static readonly XmlSerializer OrdersRelationsSerializer = new Native.Serialization.GeneratedSerializers.NWindOrderToNewEmployeeArray.ArrayOfNWindOrderToNewEmployeeSerializer();
		public static List<Employee> DataSource { get { return (List<Employee>)Serializer.Deserialize(GetDataStream()); } }
		static NWindOrderToNewEmployee[] GetOrdersRelationsList() { return (NWindOrderToNewEmployee[])OrdersRelationsSerializer.Deserialize(GetOrdersRelationsStream()); }
		static readonly Lazy<Dictionary<int, int>> ordersRelations = new Lazy<Dictionary<int, int>>(() => GetOrdersRelationsList().ToDictionary(x => x.NWindOrderId, x => x.EmployeeId));
		public static Dictionary<int, int> OrdersRelations { get { return ordersRelations.Value; } }
		public static Stream GetDataStream() { return AssemblyHelper.GetResourceStream(typeof(EmployeesWithPhotoData).Assembly, "Data/EmployeesWithPhoto.xml", true); }
		static Stream GetOrdersRelationsStream() { return AssemblyHelper.GetResourceStream(typeof(EmployeesWithPhotoData).Assembly, "Data/NWindOrdToNewEmployee.xml", true); }
	}
#pragma warning restore DX0008
	public class NWindOrderToNewEmployee {
		public int NWindOrderId { get; set; }
		public int EmployeeId { get; set; }
	}
}

File: DevExpress/XPF/DevExpress.Xpf.DemoBase/DevExpress.Xpf.DemoBase/DemosHelpers/DataClasses.vb
#If Not SILVERLIGHT Then
Imports System.Data
#End If
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.IO
Imports System.Windows.Media
Imports System.Xml.Serialization
Imports DevExpress.DemoData.Helpers
Imports DevExpress.Utils

Namespace DevExpress.Xpf.DemoBase.DataClasses
    Public Class Movie
        Public Property ID() As Integer
        Public Property Name() As String
        Public Property OrderDate() As Date
        Public Property Price() As Decimal
        Public Property Quantity() As Integer
        Public ReadOnly Property Total() As Decimal
            Get
                Return Price * Quantity
            End Get
        End Property
        Public Property Preview() As String
    End Class
    Public Class Cars

        Private imageSource_Renamed As ImageSource

        Public Property ID() As Integer
        Public Property Trademark() As String
        Public Property Model() As String
        Public Property HP() As Integer
        Public Property Liter() As Double
        Public Property Cyl() As Integer
        <XmlElement("Transmiss Speed Count")>
        Public Property TransmissSpeedCount() As Integer
        <XmlElement("Transmiss Automatic")>
        Public Property TransmissAutomatic() As String
        Public Property MPGCity() As Integer
        Public Property MPGHighway() As Integer
        Public Property Category() As String
        Public Property Description() As String
        Public Property Hyperlink() As String
        Public Property Picture() As Byte()
        Public ReadOnly Property ImageSource() As ImageSource
            Get
                If imageSource_Renamed Is Nothing Then
                    imageSource_Renamed = ImageSourceHelper.GetImageSource(New MemoryStream(Picture))
                End If
                Return imageSource_Renamed
            End Get
        End Property
        Public Property Price() As Decimal
        <XmlElement("Delivery Date")>
        Public Property DeliveryDate() As Date
        <XmlElement("Is In Stock")>
        Public Property IsInStock() As Boolean
    End Class
    Public Class Order
        Public Property ID() As Integer
        Public Property Name() As String
        Public Property OrderDate() As Date
        Public Property Price() As Decimal
        Public Property Quantity() As Integer
        Public ReadOnly Property Total() As Decimal
            Get
                Return Price * Quantity
            End Get
        End Property
        Public Property Preview() As String
    End Class
    Public Class Employee
        Implements IComparable, INotifyPropertyChanged

        Public Property Id() As Integer
        Public Property ParentId() As Integer
        Public Property FirstName() As String
        Public Property MiddleName() As String
        Public Property LastName() As String
        Public Property JobTitle() As String
        Public Property Phone() As String
        Public Property EmailAddress() As String
        Public Property AddressLine1() As String
        Public Property City() As String
        Public Property StateProvinceName() As String
        Public Property PostalCode() As String
        Public Property CountryRegionName() As String
        <XmlIgnore>
        Private groupNameCore As String
        Public Property GroupName() As String
            Get
                Return groupNameCore
            End Get
            Set(ByVal value As String)
                If groupNameCore <> value Then
                    RaisePropertyChanged("GroupName")
                    groupNameCore = value
                End If
            End Set
        End Property
        Public Property BirthDate() As Date
        Public Property HireDate() As Date
        Public Property Gender() As String
        Public Property MaritalStatus() As String
        Public Property Title() As String
        Public Property ImageData() As Byte()
        Public ReadOnly Property Image() As ImageSource
            Get
                Return If(Gender = "F", EmployeesData.ImageFemale, EmployeesData.ImageMale)
            End Get
        End Property
        Public Overrides Function ToString() As String
            Return FirstName & " " & LastName
        End Function
        #Region "Equality"
        Public Function CompareTo(ByVal obj As Object) As Integer Implements IComparable.CompareTo
            Dim empl As Employee = TryCast(obj, Employee)
            If empl Is Nothing Then
                Throw New ArgumentException()
            End If
            Return String.Compare(FirstName & LastName, empl.FirstName & empl.LastName)
        End Function
        #End Region
        Public Overrides Function Equals(ByVal obj As Object) As Boolean
            If obj Is Nothing Then
                Return False
            End If
            Return ToString() = obj.ToString()
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return Id
        End Function

        Private Sub RaisePropertyChanged(ByVal propertyName As String)
            If PropertyChangedEvent IsNot Nothing Then
                Dim e As New PropertyChangedEventArgs(propertyName)
                RaiseEvent PropertyChanged(Me, e)
            End If
        End Sub

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
    End Class
    Public Class OrderDataSourceCreator
        Public Shared Function CreateDataSource() As List(Of Order)
            Return CreateDataSource(1000)
        End Function
        Public Shared Function CreateDataSource(ByVal rowCount As Integer) As List(Of Order)
            Dim list As New List(Of Order)()
            For i As Integer = 0 To rowCount - 1
                list.Add(CreateItem(i + 1))
            Next i
            Return list
        End Function
        Private Shared Function CreateItem(ByVal id As Integer) As Order
            Dim order As New Order()
            order.ID = id
            order.Name = "Name " & id.ToString()
            order.OrderDate = Date.Today.AddDays(-id Mod 20)
            order.Price = 100 + id Mod 10
            order.Quantity = 10 + id Mod 15
            order.Preview = "c:/0.png"
            Return order
        End Function
    End Class
    Public Class MovieDataSourceCreator
        Public Shared Function CreateDataSource() As List(Of Movie)
            Return CreateDataSource(10)
        End Function
        Public Shared Function CreateDataSource(ByVal rowCount As Integer) As List(Of Movie)
            Dim list As New List(Of Movie)()
            For i As Integer = 0 To 999
                list.Add(CreateItem(i + 1))
            Next i
            Return list
        End Function
        Private Shared Function CreateItem(ByVal id As Integer) As Movie
            Dim order As New Movie()
            order.ID = id
            order.Name = "Name " & id.ToString()
            order.OrderDate = Date.Today.AddDays(-id Mod 20)
            order.Price = 100 + id Mod 10
            order.Quantity = 10 + id Mod 15
            order.Preview = "movies/0.wmv"
            Return order
        End Function
    End Class
    <XmlRoot("NewDataSet")>
    Public Class CarsData
        Inherits List(Of Cars)

        Private Shared Serializer As New XmlSerializer(GetType(CarsData))
        Public Shared ReadOnly Property NewDataSource() As List(Of Cars)
            Get
                Return DirectCast(Serializer.Deserialize(GetDataStream()), List(Of Cars))
            End Get
        End Property
        Public Shared ReadOnly Property DataSource() As List(Of Cars)
            Get
                Return NewDataSource
            End Get
        End Property
#If Not SILVERLIGHT Then
        Public Shared ReadOnly Property NewDataView() As DataView
            Get
                Dim ds As New DataSet()
                ds.ReadXml(GetDataStream())
                Return ds.Tables(0).DefaultView
            End Get
        End Property
#End If
        Private Shared Function GetDataStream() As Stream
            Return AssemblyHelper.GetResourceStream(GetType(CarsData).Assembly, "Data/Cars.xml", True)
        End Function
    End Class
    <XmlRoot("Employees")>
    Public Class EmployeesData
        Inherits List(Of Employee)

        Private Shared imageMaleSrc As ImageSource
        Private Shared imageFemaleSrc As ImageSource
        Private Shared Serializer As New XmlSerializer(GetType(EmployeesData))
        Public Shared ReadOnly Property NewDataSource() As List(Of Employee)
            Get
                Return DirectCast(Serializer.Deserialize(GetDataStream()), List(Of Employee))
            End Get
        End Property
        Public Shared ReadOnly Property DataSource() As List(Of Employee)
            Get
                Return NewDataSource
            End Get
        End Property
#If Not SILVERLIGHT Then
        Public Shared ReadOnly Property NewDataView() As DataView
            Get
                Dim ds As New DataSet()
                ds.ReadXml(GetDataStream())
                Return ds.Tables(0).DefaultView
            End Get
        End Property
#End If
        Public Shared Function GetDataStream() As Stream
            Return AssemblyHelper.GetResourceStream(GetType(EmployeesData).Assembly, "Data/Employees.xml", True)
        End Function
        Public Shared ReadOnly Property ImageMale() As ImageSource
            Get
                If imageMaleSrc Is Nothing Then
                    imageMaleSrc = ImageSourceHelper.GetImageSource(AssemblyHelper.GetResourceUri(GetType(EmployeesData).Assembly, "DemosHelpers/Images/Person_Male.png"))
                End If
                Return imageMaleSrc
            End Get
        End Property
        Public Shared ReadOnly Property ImageFemale() As ImageSource
            Get
                If imageFemaleSrc Is Nothing Then
                    imageFemaleSrc = ImageSourceHelper.GetImageSource(AssemblyHelper.GetResourceUri(GetType(EmployeesData).Assembly, "DemosHelpers/Images/Person_Female.png"))
                End If
                Return imageFemaleSrc
            End Get
        End Property
    End Class
    <XmlRoot("Employees")>
    Public Class EmployeesWithPhotoData
        Inherits List(Of Employee)

        Private Shared Serializer As New XmlSerializer(GetType(EmployeesWithPhotoData))
        Private Shared OrdersRelationsSerializer As New XmlSerializer(GetType(List(Of NWindOrderToNewEmployee)))
        Public Shared ReadOnly Property NewDataSource() As List(Of Employee)
            Get
                Return DirectCast(Serializer.Deserialize(GetDataStream()), List(Of Employee))
            End Get
        End Property
        Public Shared ReadOnly Property DataSource() As List(Of Employee)
            Get
                Return NewDataSource
            End Get
        End Property


        Private Shared ordersRelations_Renamed As List(Of NWindOrderToNewEmployee)
        Public Shared ReadOnly Property OrdersRelations() As List(Of NWindOrderToNewEmployee)
            Get
                If ordersRelations_Renamed Is Nothing Then
                    ordersRelations_Renamed = DirectCast(OrdersRelationsSerializer.Deserialize(GetOrdersRelationsStream()), List(Of NWindOrderToNewEmployee))
                End If
                Return ordersRelations_Renamed
            End Get
        End Property


        Private Shared ordersRelationsDictionary_Renamed As Dictionary(Of Integer, Integer)
        Public Shared ReadOnly Property OrdersRelationsDictionary() As Dictionary(Of Integer, Integer)
            Get
                If ordersRelationsDictionary_Renamed Is Nothing Then
                    ordersRelationsDictionary_Renamed = New Dictionary(Of Integer, Integer)()
                    For Each rel As NWindOrderToNewEmployee In OrdersRelations
                        ordersRelationsDictionary_Renamed.Add(rel.NWindOrderId, rel.EmployeeId)
                    Next rel
                End If
                Return ordersRelationsDictionary_Renamed
            End Get
        End Property

#If Not SILVERLIGHT Then
        Public Shared ReadOnly Property NewDataView() As DataView
            Get
                Dim ds As New DataSet()
                ds.ReadXml(GetDataStream())
                Return ds.Tables(0).DefaultView
            End Get
        End Property
#End If

        Public Shared Function GetDataStream() As Stream
            Return AssemblyHelper.GetResourceStream(GetType(EmployeesWithPhotoData).Assembly, "Data/EmployeesWithPhoto.xml", True)
        End Function
        Public Shared Function GetOrdersRelationsStream() As Stream
            Return AssemblyHelper.GetResourceStream(GetType(EmployeesWithPhotoData).Assembly, "Data/NWindOrdToNewEmployee.xml", True)
        End Function
    End Class

    Public Class NWindOrderToNewEmployee
        Public Property NWindOrderId() As Integer
        Public Property EmployeeId() As Integer
    End Class
End Namespace